browser/ui/webui/brave_new_tab_page_refresh/top_sites_facade.cc

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2025 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/browser/ui/webui/brave_new_tab_page_refresh/top_sites_facade.h"
+
+#include <utility>
+
+#include "base/strings/utf_string_conversions.h"
+#include "chrome/browser/ui/webui/new_tab_page/ntp_pref_names.h"
+#include "components/ntp_tiles/constants.h"
+#include "components/prefs/pref_service.h"
+
+namespace brave_new_tab_page_refresh {
+
+namespace {
+
+std::vector<mojom::TopSitePtr> TopSitesFromSections(
+    const std::map<ntp_tiles::SectionType, ntp_tiles::NTPTilesVector>&
+        sections) {
+  std::vector<mojom::TopSitePtr> top_sites;
+  for (auto& tile : sections.at(ntp_tiles::SectionType::PERSONALIZED)) {
+    auto site = mojom::TopSite::New();
+    site->title = base::UTF16ToUTF8(tile.title);
+    site->url = tile.url.spec();
+    site->favicon = tile.favicon_url.spec();
+    if (site->title.empty()) {
+      site->title = site->url;
+    }
+    top_sites.push_back(std::move(site));
+  }
+  return top_sites;
+}
+
+}  // namespace
+
+TopSitesFacade::TopSitesFacade(
+    std::unique_ptr<ntp_tiles::MostVisitedSites> most_visited_sites,
+    PrefService& pref_service)
+    : most_visited_sites_(std::move(most_visited_sites)),
+      pref_service_(pref_service) {
+  CHECK(most_visited_sites_);
+
+  pref_change_registrar_.Init(&pref_service_.get());
+  pref_change_registrar_.Add(ntp_prefs::kNtpShortcutsVisible,
+                             base::BindRepeating(&TopSitesFacade::OnPrefChanged,
+                                                 weak_factory_.GetWeakPtr()));
+  pref_change_registrar_.Add(ntp_prefs::kNtpUseMostVisitedTiles,
+                             base::BindRepeating(&TopSitesFacade::OnPrefChanged,
+                                                 weak_factory_.GetWeakPtr()));
+
+  SyncMostVisitedSites();
+  most_visited_sites_->AddMostVisitedURLsObserver(
+      this, ntp_tiles::kMaxNumMostVisited);
+}
+
+TopSitesFacade::~TopSitesFacade() = default;
+
+bool TopSitesFacade::GetTopSitesVisible() {
+  return pref_service_->GetBoolean(ntp_prefs::kNtpShortcutsVisible);
+}
+
+void TopSitesFacade::SetTopSitesVisible(bool visible) {
+  pref_service_->SetBoolean(ntp_prefs::kNtpShortcutsVisible, visible);
+  most_visited_sites_->SetShortcutsVisible(visible);
```

Comment:
Would it make more sense to set this in response to the pref change via a listener?

Reply:
Yep - makes sense.